auxiliary/proresdecoder.h

```diff
@@ -0,0 +1,148 @@
+#define PR_PIXEL_FORMAT_DEFINED
+#endif
+
+
+/*!
+ * Downscale modes. Modes smaller than full size may be used for fast,
+ * draft-mode decoding and are only supported with the '2vuy' pixel format.
+ */
+typedef enum {
+    kPRDownscale_1x1,                   ///!< Full size
+    kPRDownscale_1x2,                   ///!< Full width x 1/2 height (single-field)
+    kPRDownscale_2x2,                   ///!< 1/2 width x 1/2 height
+    kPRDownscale_4x4,                   ///!< 1/4 width x 1/4 height
+    kPRDownscale_8x8,                   ///!< 1/8 width x 1/8 height
+
+    // Old enum names for backwards compatibility
+    kPRFullSize     = kPRDownscale_1x1,
+    kPRHalfSize     = kPRDownscale_1x2,
+    kPRQuarterSize  = kPRDownscale_2x2
+
+} PRDownscaleMode;
+
+
+/*!
+ * Represents a decoder instance created by PROpenDecoder() and disposed by
+ * PRCloseDecoder().
+ */
+typedef struct PRDecoder* PRDecoderRef;
+
+
+/*!
+ * Specifies a destination buffer for the decoded image.
+ */
+struct PRPixelBuffer {
+    unsigned char*  baseAddr;       //!< Pointer to first pixel (must be 16-byte aligned)
+    int             rowBytes;       //!< Number of bytes from first pixel of one line to first pixel of next line (must be at least PRBytesPerRowNeededInPixelBuffer() and a multiple of 16)
+    PRPixelFormat   format;         //!< Pixel format (2vuy, v210, v216, etc.)
+    int             width;          //!< Buffer width in pixels
+    int             height;         //!< Buffer height in pixels
+};
+typedef struct PRPixelBuffer PRPixelBuffer;
+
+
+/*!
+ * Utility routine that returns the minimum bytes per row needed in the
+ * destination pixel buffer in order to decode a frame of the specified width
+ * using the specified pixel format and downscale mode.
+ *
+ * @param frameWidth        Full-size frame width (not the downscaled width)
+ * @param pixelFormat       Pixel format of the pixel buffer
+ * @param downscaleMode     Downscaling mode
+ */
+int
+PRBytesPerRowNeededInPixelBuffer(
+        int frameWidth,
+        PRPixelFormat pixelFormat, 
+        PRDownscaleMode downscaleMode);
+
+
+/*!
+ * Opens a decoder instance and spawns "worker" threads.  If a
+ * threadStartupCallback callback function is supplied, each worker
+ * thread will call this function upon startup to provide a way for you
+ * to adjust the threads' priorities if necessary.
+ *
+ * @param numThreads    The number of simultaneous processing threads to spawn.
+ *                      Set this to 0 to have the decoder spawn according to
+ *                      the number of processors detected in the system.
+ * @param threadStartupCallback     An optional callback function for each
+ *                                  thread to call upon startup.  Set to NULL
+ *                                  if no callback is needed.
+ *
+ * @return A reference to the instantiated decoder, or NULL on failure.
+ */
+PRDecoderRef
+PROpenDecoder(int numThreads, void (*threadStartupCallback)());
```

Comment:
*Important NOTE*
We haven't yet explored the `numThreads` opportunity provided by the decoder itself. 
We are only use the FFmpeg threading per frame feature.